### PR TITLE
add crop transforms with unit tests for lazy resampling

### PR DESCRIPTION
**Pull request by KumoLiu**
Part of #5991 .

### Description

This PR adds the crop transforms from https://github.com/Project-MONAI/MONAI/pull/5860 relate to lazy resampling, and it belongs to the second part mentioned in https://github.com/Project-MONAI/MONAI/issues/5991

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: tests/test_center_scale_crop.py [hunk 1]
```diff
@@ -20,7 +20,9 @@
 from tests.croppers import CropTest
 
 TEST_SHAPES = [
-    [{"roi_scale": [0.6, 0.3, -1]}, (3, 3, 3, 3), (3, 2, 1, 3)],
+    # TODO issue with single-dimensional output in all crop/pad transform
+    # [{"roi_scale": [0.6, 0.3, -1]}, (3, 3, 3, 3), (3, 2, 1, 3)],
+    [{"roi_scale": [0.6, 0.3, -1]}, (3, 3, 4, 3), (3, 2, 2, 3)],
```
[the remark this answers is not in the document]
Comment (wyli) [reply tests/test_center_scale_crop.py:25]: set `align_corners=False` in `apply_transforms`  may close the gaps in some cases

[the remark this answers is not in the document]
Comment (KumoLiu) [reply tests/test_center_scale_crop.py]: Thanks @wyli for the solution, have fixed it in this [commit](https://github.com/Project-MONAI/MONAI/pull/6068/commits/ad45cfe100141942240b66a43a6ab02ed2b1ec75).

File: tests/test_spatial_crop.py [hunk 2]
```diff
@@ -25,7 +25,7 @@
     [{"roi_start": [0, 0], "roi_end": [2, 2]}, (3, 3, 3, 3), (3, 2, 2, 3)],
     [{"roi_start": [0, 0, 0, 0, 0], "roi_end": [2, 2, 2, 2, 2]}, (3, 3, 3, 3), (3, 2, 2, 2)],
     [{"roi_start": [0, 0, 0, 0, 0], "roi_end": [8, 8, 8, 2, 2]}, (3, 3, 3, 3), (3, 3, 3, 3)],
-    [{"roi_start": [1, 0, 0], "roi_end": [1, 8, 8]}, (3, 3, 3, 3), (3, 0, 3, 3)],
+    # [{"roi_start": [1, 0, 0], "roi_end": [1, 8, 8]}, (3, 3, 3, 3), (3, 0, 3, 3)], # not support during lazy
```
Comment (KumoLiu) [inline tests/test_spatial_crop.py:28]: When output data contains zero dimension(None), it's not supported in the lazy mode. Even after changing the `func` with `x>=0`, we can get the correct `spatial_size` [here](https://github.com/Project-MONAI/MONAI/blob/26f100aaf27cc9e4fd404d6ec8271b2c5e2ad8f2/monai/transforms/spatial/functional.py#L98), but zero is still not supported by `nn.functional.affine_grid`.
https://github.com/Project-MONAI/MONAI/blob/f519699f4b709aa69f23c252842e4798cac293d8/monai/utils/misc.py#L195
https://github.com/Project-MONAI/MONAI/blob/26f100aaf27cc9e4fd404d6ec8271b2c5e2ad8f2/monai/transforms/spatial/functional.py#L98

Comment (wyli) [reply tests/test_spatial_crop.py:28]: sure, that's expected, I think that's an ill-pose case, raising an error is fine..

Comment (KumoLiu) [reply tests/test_spatial_crop.py:28]: ok, will add it.

Comment (KumoLiu) [reply tests/test_spatial_crop.py]: updated in this [commit](https://github.com/Project-MONAI/MONAI/pull/6068/commits/4a7b4bac7185ad6709f7f71bdfa52378191589cf).

**Comment by KumoLiu** [comment]
Just set this PR to be ready for review. And will create another pr for `CropForeground` and other random crop transforms to avoid large PR.

**Review by wyli — APPROVED** [review]

File: monai/transforms/croppad/functional.py [hunk 3]
```diff
@@ -129,3 +129,40 @@ def pad_func(img: torch.Tensor, to_pad: list[tuple[int, int]], mode: str, transf
...
+    shape = [s.indices(o)[1] - s.indices(o)[0] for s, o in zip(slices[1:], img_size)]
+    meta_info = TraceableTransform.track_transform_meta(
+        img,
+        sp_size=shape,
+        affine=create_translate(spatial_rank, to_shift),
+        extra_info=extra_info,
+        orig_size=img_size,
+        transform_info=transform_info,
+        lazy_evaluation=transform_info.get(TraceKeys.LAZY_EVALUATION, False),
+    )
```
Comment (Nic-Ma) [inline monai/transforms/croppad/functional.py:163]: Hi @wyli @KumoLiu ,

In the `pad_func`, we have a function `pad_nd` as public API:
```py
out = pad_nd(out, to_pad, mode, **kwargs) if do_pad else out
out = convert_to_tensor(out, track_meta=get_track_meta())
```
Is it better to also abstract a `crop_nd` here?

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Overall looks good to me, put some minor coments inline.

Thanks.

File: monai/transforms/croppad/functional.py (hunk 3 above)
Comment (wyli) [reply monai/transforms/croppad/functional.py:163]: Integer slicing is quite straightforward for both array and tensor... img[slices] I think we can add it when we consider non-integer indexing, we currently don't have this implementation

**Pull request merged**